Content.Server/Fluids/EntitySystems/SpillableSystem.cs

```diff
@@ -1,33 +1,196 @@
 using Content.Server.Chemistry.EntitySystems;
-using Content.Server.Construction.Components;
+using Content.Server.Coordinates.Helpers;
 using Content.Server.Fluids.Components;
-using Content.Shared.Examine;
+using Content.Shared.Administration.Logs;
+using Content.Shared.Chemistry.Components;
+using Content.Shared.Chemistry.Reagent;
+using Content.Shared.Database;
+using Content.Shared.FixedPoint;
 using Content.Shared.Throwing;
 using Content.Shared.Verbs;
 using JetBrains.Annotations;
+using Robust.Server.GameObjects;
 using Robust.Shared.GameObjects;
 using Robust.Shared.IoC;
+using Robust.Shared.Localization;
+using Robust.Shared.Map;
+using Robust.Shared.Prototypes;
 
 namespace Content.Server.Fluids.EntitySystems;
 
 [UsedImplicitly]
 public class SpillableSystem : EntitySystem
 {
     [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+    [Dependency] private readonly PuddleSystem _puddleSystem = default!;
+    [Dependency] private readonly IMapManager _mapManager = default!;
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+    [Dependency] private readonly IEntityLookup _entityLookup = default!;
+    [Dependency] private readonly GridTileLookupSystem _gridTileLookupSystem = default!;
 
-    public override void Initialize() 
+    public override void Initialize()
     {
         base.Initialize();
         SubscribeLocalEvent<SpillableComponent, LandEvent>(SpillOnLand);
+        SubscribeLocalEvent<SpillableComponent, GetOtherVerbsEvent>(AddSpillVerb);
     }
 
-    void SpillOnLand(EntityUid uid, SpillableComponent component, LandEvent args) {
-        if (args.User != null && _solutionContainerSystem.TryGetSolution(uid, component.SolutionName, out var solutionComponent))
+    /// <summary>
+    ///     Spills the specified solution at the entity's location if possible.
+    /// </summary>
+    /// <param name="uid">
+    ///     The entity to use as a location to spill the solution at.
+    /// </param>
+    /// <param name="solution">Initial solution for the prototype.</param>
+    /// <param name="prototype">The prototype to use.</param>
+    /// <param name="sound">Play the spill sound.</param>
+    /// <param name="combine">Whether to attempt to merge with existing puddles</param>
+    /// <param name="transformComponent">Optional Transform component</param>
+    /// <returns>The puddle if one was created, null otherwise.</returns>
+    public PuddleComponent? SpillAt(EntityUid uid, Solution solution, string prototype,
+        bool sound = true, bool combine = true, TransformComponent? transformComponent = null)
+    {
+        return !Resolve(uid, ref transformComponent, false)
+            ? null
+            : SpillAt(solution, transformComponent.Coordinates, prototype, sound: sound, combine: combine);
+    }
+
+    private void SpillOnLand(EntityUid uid, SpillableComponent component, LandEvent args)
+    {
+        if (args.User != null &&
```

Comment:
It was like this before, but why is this checking if `user == null`? AFAICT, this means that if an entity is not thrown by a user (e.g., space wind or explosions), then this will not spill. Is that intentional?

Reply:
Also, it might be worth adding an admin log here. There are already `LogType.Landed` logs with information about the entity & thrower, so I guess just adding a "[entity] spilled [quantity] reagent on landing" log would probably be enough information. I don't think there is an easy way to summarize the contents of a solution for log?. 

The spill verb will already create a log (although its a generic verb one, with no reagent information).